domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,376 @@
+//
+// Custom images doesn't expire.
+func (s *State) tryCleanUpExpiredMetadata(ctx context.Context) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	deleteQuery := fmt.Sprintf(`
+DELETE FROM cloud_image_metadata 
+WHERE source != '%s' AND  
+created_at < '%s';`, CustomSource, time.Now().Add(-ExpirationDelay).Format(time.DateTime))
+	deleteMetadataStmt, err := sqlair.Prepare(deleteQuery)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, deleteMetadataStmt).Run()
+		return err
+	}))
+}
+
+// ensuresArchitecturesLoaded initializes architecture data by loading it from the database if not already done.
+// architecture are readonly, inserted at the creation of the DB, so it should be ok to load them once at the start
+// of the controller.
+func (s *State) ensuresArchitecturesLoaded(ctx context.Context) (err error) {
+	s.archesOnce.Do(func() {
+		defer func() {
+			if err != nil {
+				s.archesOnce = sync.Once{} // reset Once since something has been wrong during the init process
+				s.architectureIDsByName = nil
+			}
+		}()
+
+		var db database.TxnRunner
+		db, err = s.DB()
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		var loadArchsStmt *sqlair.Statement
+		loadArchsStmt, err = sqlair.Prepare(`SELECT &architecture.* FROM architecture;`, architecture{})
+
+		var archs []architecture
+		err = errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+			return tx.Query(ctx, loadArchsStmt).GetAll(&archs)
+		}))
+		if err != nil {
+			return
+		}
+
+		s.architectureIDsByName = make(map[string]int, len(archs))
+		for _, arch := range archs {
+			s.architectureIDsByName[arch.Name] = arch.ID
+		}
+	})
+
+	return err
+}
+
+// validateMetadata checks the required fields in the provided Metadata and returns an error if any field is missing.
+func validateMetadata(m cloudimagemetadata.Metadata) error {
+	if m.Version == "" {
+		return errors.Annotatef(cloudmetadataerrors.Invalid, "missing version: metadata for image %v", m.ImageId)
+	}
+	if m.Stream == "" {
+		return errors.Annotatef(cloudmetadataerrors.Invalid, "missing stream: metadata for image %v", m.ImageId)
+	}
+	if m.Source == "" {
+		return errors.Annotatef(cloudmetadataerrors.Invalid, "missing source: metadata for image %v", m.ImageId)
+	}
+	if m.Arch == "" {
+		return errors.Annotatef(cloudmetadataerrors.Invalid, "missing architecture: metadata for image %v", m.ImageId)
+	}
+	if m.Region == "" {
+		return errors.Annotatef(cloudmetadataerrors.Invalid, "missing region: metadata for image %v", m.ImageId)
+	}
+	return nil
+}
+
+// buildKey constructs a unique key string by concatenating various metadata attributes.
+func buildKey(m cloudimagemetadata.Metadata) string {
+	return fmt.Sprintf("%s:%s:%s:%s:%s:%s:%s",
+		m.Stream,
+		m.Region,
+		m.Version,
+		m.Arch,
+		m.VirtType,
+		m.RootStorageType,
+		m.Source)
+}
```

Comment:
OK, I've just seen the current code, and this is asinine.

https://github.com/juju/juju/blob/25923e1c98a2443adfad4f02e8babfa029f5079d/state/cloudimagemetadata/image.go#L299-L308

Reply:
I vote we give it a UUID and then you can search for a cloud image by those fields. It will then return you a Metadata based on that.

```go
func (s *Service) GetImageMetadataUUID(ctx context.Context, metadataFields MetadataFields) (UUID, error) {
```

Then all service methods should take a UUID.

Reply:
That's fun how you point out those point where I walk on it and don't find a nice solution.

To be honest, I hesitated between this solution and a big "WHERE clause". I finally keep the old way because it's simpler, and most of the data in this base is only cached from simple stream.

However, since I am not the only which feels it wrong, this is my alternate solution I think of before keeping the "good old way":

- Add a generated uuid instead of the key (possibly generated from sqlite)
- Gather updatable items through a big WHERE clause
- Use the uuid to update only those which are actually concerned.
- The strangee multiple insert error could be only raised to translate a primary key constraint error.

It is slightly more complicated, it's why i didn't keep it.

Reply:
I'm not convinced to expose UUID here. It would cause very tricky question like:

- FindMetada(someCriteria) => get a bunch of unexpired UUID
- GetMetadataBYUUID(myUUID) => what if my UUID has expired in between ? Do I return it ? If yes it may also have been deleted... So we have a three different possible outcome, and 2/3 are error-prone.

The actual interface is better until we don't change the whole stack (as we discussed).

=> have a true cache for simplestream metadata
=> have a store (with UUID-alike interface) for custom metadata.

For now, we have a cachestore metadata stuff weirdness, I am not very confident to not stick to the old interface...

Reply:
It doesn't matter if it's expired in between the two calls, just return a NotFound error. There is no downside. You made one more request in the low read/write collection. I'm not going to break a sweat over it.

To wit, it has been decided that we should work on UUIDs for all service methods with the exception of getting the UUID using the natural key. This way we can push the UUID up the stack and eventually into the API server layer. @manadart can confirm.

Reply:
Yes I agree about using UUID, generally speaking for all service methods regarding **resources** (ie, machine, units, models, whatever).

But, as per my comprehension of what are CloudMetadata, they can't be considered as **resources**, for several reason:

* they don't have a "natural key" => they can't have a public uuid either.
* they are (may, in fact) volatile data (usually a resources isn't that volatile.
* there is no such method as "GetMetadata", only a "FindMetadata" => we *always* get those by pack, over criteria.
* even "DeleteMetadata", which use an id, may delete several metadata (several metadata can share the same id for several stream/arch/region/.../whatever)

This domain doesn't modelize a bunch of resources but an index table for external resources (images, identified by their ID, handled externally, and indexed by their Metadata). So those metadata are mostly indexes, not resources, and shouldn't be handled like they were resources.

Having a very small state layer working as a simple CRUD and delegating search in service layer would lead to very boiler-plate code, and probably error-prone (at least from my point of view)

I would be happy to discuss it with you to close this point, especially if i get something wrong.

As references, there is the state interface of this service:
https://github.com/juju/juju/blob/bd1e949c27ef625ff832afdc48d99015611b2ea2/state/cloudimagemetadata/interface.go#L56-L76

Reply:
Will change accordingly our discussion:

* Remove build_key related code in favor of an UUID fetch
* Rename `DeleteMetadata` as `DeleteMetadataWithImageId` to make room to be able to add a method that delete metadata by UUID in a consistent way (either we may consistently name it `DeleteMetadataByUUID` or just `DeleteMetadata(UUID)`